### PR TITLE
Adding Account, Landing and Shared engines

### PR DESCRIPTION
**Pull request by enricostano**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -10,7 +10,9 @@ gem 'bcrypt-ruby', '~> 3.1.5'
 
 gem 'pundit'
 
-gem 'public_pages', path: "engines/public_pages"
+gemspec path: "engines/shared"
+gemspec path: "engines/account"
+gemspec path: "engines/landing"
```
Comment (andresgutgon) [inline Gemfile]: Are you adding landing engine here?

File: config/routes.rb [hunk 2]
```diff
@@ -1,23 +1,5 @@
...
-      end
-      resources :invitations, except: [:new, :edit] do
-        member do
-          put :accept
-        end
-      end
-    end
-  end
+  mount Account::Engine,    at: '/'
+  mount Landing::Engine,    at: '/'
```
Comment (andresgutgon) [inline config/routes.rb]: `Shared` engine doesn't need to load nothing here? and again. Are you adding `Landing` engine in this PR?

File: db/schema.rb [hunk 3]
```diff
@@ -11,71 +11,26 @@
...
-
-  create_table "users_units", force: true do |t|
-    t.integer  "group_id",   null: false
-    t.string   "name",       null: false
-    t.datetime "created_at"
-    t.datetime "updated_at"
-  end
-
-  add_index "users_units", ["group_id"], name: "index_users_units_on_group_id", using: :btree
-
```
Comment (andresgutgon) [inline db/schema.rb]: Slim schema :trollface:

File: engines/account/MIT-LICENSE [hunk 4]
```diff
@@ -0,0 +1,20 @@
+Copyright 2015 YOURNAME
```
Comment (andresgutgon) [inline engines/account/MIT-LICENSE]: `YOURNAME`

File: engines/account/account.gemspec [hunk 5]
```diff
@@ -0,0 +1,22 @@
+$:.push File.expand_path("../lib", __FILE__)
+
+# Maintain your gem's version:
+require "account/version"
+
+# Describe your gem and declare its dependencies:
+Gem::Specification.new do |s|
+  s.name        = "account"
+  s.version     = Account::VERSION
+  s.authors     = ["TODO: Your name"]
```
Comment (andresgutgon) [inline engines/account/account.gemspec]: `authors`

File: engines/account/account.gemspec [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  s.email       = ["TODO: Your email"]
```
Comment (andresgutgon) [inline engines/account/account.gemspec]: `contact email`

File: engines/account/account.gemspec [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  s.homepage    = "TODO"
```
Comment (andresgutgon) [inline engines/account/account.gemspec]: `katuma.org`

File: engines/account/app/mailers/account/signup_mailer.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  class SignupMailer < ActionMailer::Base
+
+    default from: 'info@katuma.org'
+
+    # @param signup [Account::Signup]
+    def confirm_email(signup)
+      @signup = signup
+      @url = confirm_url
+
+      mail to: @signup.email, subject: "Welcome to Katuma!"
```
Comment (andresgutgon) [inline engines/account/app/mailers/account/signup_mailer.rb]: `i18n`. You could start adding keys to translations. Start a ES or CAT file whatever default language is for the project

File: engines/account/app/mailers/account/signup_mailer.rb [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    end
+
+    private
+
+    def confirm_url
+      complete_signup_url(@signup, host: 'localhost:3000', email: @signup.email, token: @signup.token)
```
Comment (andresgutgon) [inline engines/account/app/mailers/account/signup_mailer.rb]: MEG :warning:

File: engines/account/app/models/account/group.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+module Account
+  class Group < ActiveRecord::Base
+
+    self.table_name = :groups
```
Comment (andresgutgon) [inline engines/account/app/models/account/group.rb]: Is this used in this iteration?

File: engines/account/app/models/account/membership.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+module Account
+  class Membership < ActiveRecord::Base
+
+    self.table_name = :memberships
+
+    include Shared::Model::ReadOnly
+
+    belongs_to :user
+    belongs_to :group
```
Comment (andresgutgon) [inline engines/account/app/models/account/membership.rb]: ditto

File: engines/account/app/models/account/signup.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+    self.table_name = :signups
+
+    before_validation do |signup|
+      signup.token = SecureRandom.urlsafe_base64(32)
+    end
+
+    validates :email,
+      presence: true
+    validates :token,
```
Comment (andresgutgon) [inline engines/account/app/models/account/signup.rb]: One line is cleaner IMO

File: engines/account/app/models/account/user.rb [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+    self.table_name = :users
+
+    has_secure_password
+
+    validates :name,
+      presence: true
+    validates :email,
+      presence: true,
+      uniqueness: true
```
Comment (andresgutgon) [inline engines/account/app/models/account/user.rb]: One line maybe

File: engines/account/app/views/account/session/login.html.erb [hunk 14]
```diff
@@ -0,0 +1,31 @@
+<% if flash[:error] %>
+  <p><%= flash[:error] %></p>
+<% end %>
```
Comment (andresgutgon) [inline engines/account/app/views/account/session/login.html.erb]: Flash could go on the layout not in every template

File: engines/account/app/views/account/signup/complete.html.erb [hunk 15]
```diff
@@ -0,0 +1,37 @@
+<% if flash[:error] %>
+  <p><%= flash[:error] %></p>
+<% end %>
```
Comment (andresgutgon) [inline engines/account/app/views/account/signup/complete.html.erb]: Ditto. Not for now, but I think you could change

File: engines/account/spec/dummy/app/views/layouts/application.html.erb [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+  <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track' => true %>
+  <%= javascript_include_tag 'application', 'data-turbolinks-track' => true %>
+  <%= csrf_meta_tags %>
+</head>
+<body>
+
+<%= yield %>
+
+</body>
+</html>
```
Comment (andresgutgon) [inline engines/account/spec/dummy/app/views/layouts/application.html.erb]: If you're not going to do nothing with `dummy` app in this iteration or next. You could get rid of it and add when needed. Less :hankey:

**Comment by andresgutgon** [comment]
:postal_horn: :green_apple: :+1:

File: Gemfile (hunk 1 above)
Comment (enricostano) [inline Gemfile]: Yep, it's just the landing page where you go if you're not logged in

File: config/routes.rb (hunk 2 above)
Comment (enricostano) [inline config/routes.rb]: Shared only provides mixins and classes to be used in other engines, no routes here.

Ditto for the Landing engine

File: engines/account/app/mailers/account/signup_mailer.rb (hunk 8 above)
Comment (enricostano) [inline engines/account/app/mailers/account/signup_mailer.rb]: yep that's something that we need ASAP of course, maybe last thing after all the behaviour works as expected

File: engines/account/app/mailers/account/signup_mailer.rb (hunk 9 above)
Comment (enricostano) [inline engines/account/app/mailers/account/signup_mailer.rb]: I know, will move this to a config ASAP

File: engines/account/app/models/account/group.rb (hunk 10 above)
Comment (enricostano) [inline engines/account/app/models/account/group.rb]: ooops :grin:

File: engines/account/app/models/account/signup.rb (hunk 12 above)
Comment (enricostano) [inline engines/account/app/models/account/signup.rb]: We can use multiple lines when we have more than one option

File: engines/account/app/views/account/session/login.html.erb (hunk 14 above)
Comment (enricostano) [inline engines/account/app/views/account/session/login.html.erb]: sure

File: engines/account/spec/dummy/app/views/layouts/application.html.erb (hunk 16 above)
Comment (enricostano) [inline engines/account/spec/dummy/app/views/layouts/application.html.erb]: Sure, I'll do that today.

**Comment by enricostano** [comment]
Many thanks @andresgutgon !

File: engines/account/app/models/account/signup.rb (hunk 12 above)
Comment (andresgutgon) [inline engines/account/app/models/account/signup.rb]: To this `validate` fits in one line.

**Pull request merged**